staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,337 @@
+      $Loop = $false
+    }
+    catch {
+      if ($RetryCount -gt $RetryCountMax) {
+        $Loop = $false
+        throw [NotSupportedException] "Failed to retrieve VM metadata after $RetryCountMax attempts. Exiting! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+      }
+      else {
+        Start-Sleep -Seconds ($RetryCount * 2 + 1)
+        $RetryCount++
+      }
+    }
+  } while ($Loop)
+
+  return $Compute
+}
+
+function Get-DriverFile {
+  param(
+    [Parameter(Mandatory = $True)]
+    [ValidateNotNullOrEmpty()]
+    [string] $source,
+  
+    [Parameter(Mandatory = $True)]
+    [ValidateNotNullOrEmpty()]
+    [string] $dest
+  )
+  
+  function GetUsingWebClient {
+    # Store Security Protocols
+    $protocols = [Net.ServicePointManager]::SecurityProtocol
+    # Add Tls12 to Security Protocols
+    [Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12)
+
+    $wc = New-Object System.Net.WebClient
+    $start_time = Get-Date
+    $wc.DownloadFile($source, $dest)
+    Write-Host "Time taken: $((Get-Date).Subtract($start_time).Seconds) second(s)"
+
+    # Reset Security Protocols
+    [Net.ServicePointManager]::SecurityProtocol = $protocols
+  }
+  
+  function GetUsingInvokeWebRequest {
+    $start_time = Get-Date
+    Invoke-WebRequest -Uri $source -OutFile $dest
+    Write-Host "Time taken: $((Get-Date).Subtract($start_time).Seconds) second(s)"
+  }
+  
+  # Invoke-WebRequest seems slower for large files.
+  # Using System.Net/WebClient instead. Apparently wget (and csource) are aliases
+
+  Write-Host "Downloading from $source to $dest"
+
+  # Retry to overcome failure in downloading
+  $Loop = $true
+  $RetryCount = 0
+  $RetryCountMax = 10
+  do {
+    try {
+      GetUsingWebClient
+      $Loop = $false
+    }
+    catch {
+      if ($RetryCount -gt $RetryCountMax) {
+        $Loop = $false
+        throw [NotSupportedException] "Failed to download $source after $RetryCountMax attempts. Exiting! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+      }
+      else {
+        Start-Sleep -Seconds ($RetryCount * 2 + 1)
+        $RetryCount++
+      }
+    }
+  } while ($Loop)
+}
+
+function Add-DriverCertificate {
+  param(
+    [Parameter(Mandatory = $True)]
+    [ValidateNotNullOrEmpty()]
+    [string] $link
+  )
+
+  $Cert = Get-DriverCertificate $link
+  if ( $Cert.OldObject ) {
+    Write-Host 'Certificate already in store.'
+  }
+  else {
+    Write-Host 'Adding Certificate ...'
+    $Store = Get-Item $Cert.Store
+    $Store.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite)
+    Import-Certificate -Filepath $Cert.File -CertStoreLocation $Cert.Store
```

Comment:
Should we validate the result?

Reply:
It turns out the cert file is deprecated. The signature is embedded in the installer exe binary.